File: push/src/main/java/com/xiaomi/channel/commonutils/reflect/JavaCalls.java
package com.xiaomi.channel.commonutils.reflect;

import android.util.Log;
import com.xiaomi.channel.commonutils.android.SystemUtils;
import java.lang.reflect.Constructor;
import java.lang.reflect.Field;
import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.util.HashMap;
import java.util.Map;

/* JADX INFO: loaded from: miuipushsdkshared_3_7_9.jar:com/xiaomi/channel/commonutils/reflect/JavaCalls.class */
public class JavaCalls {
    private static final String LOG_TAG = "JavaCalls";
    private static final Map<Class<?>, Class<?>> PRIMITIVE_MAP;

    /* JADX INFO: loaded from: miuipushsdkshared_3_7_9.jar:com/xiaomi/channel/commonutils/reflect/JavaCalls$JavaParam.class */
    public static class JavaParam<T> {
        public final Class<? extends T> clazz;
        public final T obj;

        public JavaParam(Class<? extends T> cls, T t) {
            this.clazz = cls;
            this.obj = t;
        }
    }

    static {
        HashMap map = new HashMap();
        PRIMITIVE_MAP = map;
        map.put(Boolean.class, Boolean.TYPE);
        map.put(Byte.class, Byte.TYPE);
        map.put(Character.class, Character.TYPE);
        map.put(Short.class, Short.TYPE);
        map.put(Integer.class, Integer.TYPE);
        map.put(Float.class, Float.TYPE);
        map.put(Long.class, Long.TYPE);
        map.put(Double.class, Double.TYPE);
        map.put(Boolean.TYPE, Boolean.TYPE);
        map.put(Byte.TYPE, Byte.TYPE);
        map.put(Character.TYPE, Character.TYPE);
        map.put(Short.TYPE, Short.TYPE);
        map.put(Integer.TYPE, Integer.TYPE);
        map.put(Float.TYPE, Float.TYPE);
        map.put(Long.TYPE, Long.TYPE);
        map.put(Double.TYPE, Double.TYPE);
    }

    public static <T> T callMethod(Object obj, String str, Object... objArr) {
        try {
            return (T) callMethodOrThrow(obj, str, objArr);
        } catch (Exception e) {
            Log.w(LOG_TAG, "Meet exception when call Method '" + str + "' in " + obj + ", " + e);
            return null;
        }
    }

    public static <T> T callMethodOrThrow(Object obj, String str, Object... objArr) throws IllegalAccessException, NoSuchMethodException, SecurityException, IllegalArgumentException, InvocationTargetException {
        return (T) getDeclaredMethod(obj.getClass(), str, getParameterTypes(objArr)).invoke(obj, getParameters(objArr));
    }

    public static <T> T callStaticMethod(String str, String str2, Object... objArr) {
        try {
            return (T) callStaticMethodOrThrow(SystemUtils.loadClass(null, str), str2, objArr);
        } catch (Exception e) {
            Log.w(LOG_TAG, "Meet exception when call Method '" + str2 + "' in " + str + ", " + e);
            return null;
        }
    }

    public static <T> T callStaticMethodOrThrow(Class<?> cls, String str, Object... objArr) throws IllegalAccessException, NoSuchMethodException, SecurityException, IllegalArgumentException, InvocationTargetException {
        return (T) getDeclaredMethod(cls, str, getParameterTypes(objArr)).invoke(null, getParameters(objArr));
    }

    public static <T> T callStaticMethodOrThrow(String str, String str2, Object... objArr) throws IllegalAccessException, NoSuchMethodException, SecurityException, ClassNotFoundException, IllegalArgumentException, InvocationTargetException {
        return (T) getDeclaredMethod(SystemUtils.loadClass(null, str), str2, getParameterTypes(objArr)).invoke(null, getParameters(objArr));
    }

    private static boolean compareClassLists(Class<?>[] clsArr, Class<?>[] clsArr2) {
        if (clsArr == null) {
            return clsArr2 == null || clsArr2.length == 0;
        }
        if (clsArr2 == null) {
            return clsArr.length == 0;
        }
        if (clsArr.length != clsArr2.length) {
            return false;
        }
        for (int i = 0; i < clsArr.length; i++) {
            if (clsArr2[i] != null && !clsArr[i].isAssignableFrom(clsArr2[i])) {
                Map<Class<?>, Class<?>> map = PRIMITIVE_MAP;
                if (!map.containsKey(clsArr[i]) || !map.get(clsArr[i]).equals(map.get(clsArr2[i]))) {
                    return false;
                }
            }
        }
        return true;
    }

    private static Method findMethodByName(Method[] methodArr, String str, Class<?>[] clsArr) {
        if (str == null) {
            throw new NullPointerException("Method name must not be null.");
        }
        for (Method method : methodArr) {
            if (method.getName().equals(str) && compareClassLists(method.getParameterTypes(), clsArr)) {
                return method;
            }
        }
        return null;
    }

    private static Method getDeclaredMethod(Class<?> cls, String str, Class<?>... clsArr) throws NoSuchMethodException, SecurityException {
        Method methodFindMethodByName = findMethodByName(cls.getDeclaredMethods(), str, clsArr);
        if (methodFindMethodByName != null) {
            methodFindMethodByName.setAccessible(true);
            return methodFindMethodByName;
        }
        if (cls.getSuperclass() != null) {
            return getDeclaredMethod(cls.getSuperclass(), str, clsArr);
        }
        throw new NoSuchMethodException();
    }

    private static Object getDefaultValue(Class<?> cls) {
        if (Integer.TYPE.equals(cls) || Integer.class.equals(cls) || Byte.TYPE.equals(cls) || Byte.class.equals(cls) || Short.TYPE.equals(cls) || Short.class.equals(cls) || Long.TYPE.equals(cls) || Long.class.equals(cls) || Double.TYPE.equals(cls) || Double.class.equals(cls) || Float.TYPE.equals(cls) || Float.class.equals(cls)) {
            return 0;
        }
        if (Boolean.TYPE.equals(cls) || Boolean.class.equals(cls)) {
            return false;
        }
        return (Character.TYPE.equals(cls) || Character.class.equals(cls)) ? (char) 0 : null;
    }

    public static <T> T getField(Object obj, String str) {
        try {
            return (T) getFieldOrThrow(obj.getClass(), obj, str);
        } catch (Exception e) {
            Log.w(LOG_TAG, "Meet exception when call getField '" + str + "' in " + obj + ", " + e);
            return null;
        }
    }

    public static <T> T getFieldOrThrow(Class<? extends Object> cls, Object obj, String str) throws IllegalAccessException, NoSuchFieldException {
        Class<? extends Object> superclass = cls;
        Field field = null;
        while (field == null) {
            try {
                Field declaredField = superclass.getDeclaredField(str);
                declaredField.setAccessible(true);
                field = declaredField;
            } catch (NoSuchFieldException e) {
                superclass = superclass.getSuperclass();
            }
            if (superclass == null) {
                throw new NoSuchFieldException();
            }
        }
        field.setAccessible(true);
        return (T) field.get(obj);
    }

    private static Class<?>[] getParameterTypes(Object... objArr) {
        Class<?>[] clsArr = null;
        if (objArr != null) {
            clsArr = null;
            if (objArr.length > 0) {
                Class<?>[] clsArr2 = new Class[objArr.length];
                int i = 0;
                while (true) {
                    clsArr = clsArr2;
                    if (i >= objArr.length) {
                        break;
                    }
                    Object obj = objArr[i];
                    if (obj == null || !(obj instanceof JavaParam)) {
                        clsArr2[i] = obj == null ? null : obj.getClass();
                    } else {
                        clsArr2[i] = ((JavaParam) obj).clazz;
                    }
                    i++;
                }
            }
        }
        return clsArr;
    }

    private static Object[] getParameters(Object... objArr) {
        Object[] objArr2 = null;
        if (objArr != null) {
            objArr2 = null;
            if (objArr.length > 0) {
                Object[] objArr3 = new Object[objArr.length];
                int i = 0;
                while (true) {
                    objArr2 = objArr3;
                    if (i >= objArr.length) {
                        break;
                    }
                    Object obj = objArr[i];
                    if (obj == null || !(obj instanceof JavaParam)) {
                        objArr3[i] = obj;
                    } else {
                        objArr3[i] = ((JavaParam) obj).obj;
                    }
                    i++;
                }
            }
        }
        return objArr2;
    }

    public static <T> T getStaticField(Class<? extends Object> cls, String str) {
        try {
            return (T) getFieldOrThrow(cls, null, str);
        } catch (Exception e) {
            StringBuilder sb = new StringBuilder();
            sb.append("Meet exception when call getStaticField '");
            sb.append(str);
            sb.append("' in ");
            sb.append(cls != null ? cls.getSimpleName() : "");
            sb.append(", ");
            sb.append(e);
            Log.w(LOG_TAG, sb.toString());
            return null;
        }
    }

    public static <T> T getStaticField(String str, String str2) {
        try {
            return (T) getFieldOrThrow(SystemUtils.loadClass(null, str), null, str2);
        } catch (Exception e) {
            Log.w(LOG_TAG, "Meet exception when call getStaticField '" + str2 + "' in " + str + ", " + e);
            return null;
        }
    }

    public static <T> T newEmptyInstance(Class<?> cls) {
        try {
            return (T) newEmptyInstanceOrThrow(cls);
        } catch (Exception e) {
            Log.w(LOG_TAG, "Meet exception when make instance as a " + cls.getSimpleName() + ", " + e);
            return null;
        }
    }

    public static <T> T newEmptyInstanceOrThrow(Class<?> cls) throws IllegalAccessException, InstantiationException, ClassNotFoundException, InvocationTargetException {
        Constructor<?>[] declaredConstructors = cls.getDeclaredConstructors();
        if (declaredConstructors == null || declaredConstructors.length == 0) {
            throw new IllegalArgumentException("Can't get even one available constructor for " + cls);
        }
        Constructor<?> constructor = declaredConstructors[0];
        constructor.setAccessible(true);
        Class<?>[] parameterTypes = constructor.getParameterTypes();
        if (parameterTypes == null || parameterTypes.length == 0) {
            return (T) constructor.newInstance(new Object[0]);
        }
        Object[] objArr = new Object[parameterTypes.length];
        for (int i = 0; i < parameterTypes.length; i++) {
            objArr[i] = getDefaultValue(parameterTypes[i]);
        }
        return (T) constructor.newInstance(objArr);
    }

    public static <T> T newInstance(Class<?> cls, Object... objArr) {
        try {
            return (T) newInstanceOrThrow(cls, objArr);
        } catch (Exception e) {
            Log.w(LOG_TAG, "Meet exception when make instance as a " + cls.getSimpleName() + ", " + e);
            return null;
        }
    }

    public static Object newInstance(String str, Object... objArr) {
        try {
            return newInstanceOrThrow(str, objArr);
        } catch (Exception e) {
            Log.w(LOG_TAG, "Meet exception when make instance as a " + str + ", " + e);
            return null;
        }
    }

    public static <T> T newInstanceOrThrow(Class<?> cls, Object... objArr) throws IllegalAccessException, NoSuchMethodException, InstantiationException, SecurityException, IllegalArgumentException, InvocationTargetException {
        return (T) cls.getConstructor(getParameterTypes(objArr)).newInstance(getParameters(objArr));
    }

    public static Object newInstanceOrThrow(String str, Object... objArr) throws IllegalAccessException, NoSuchMethodException, InstantiationException, SecurityException, ClassNotFoundException, IllegalArgumentException, InvocationTargetException {
        return newInstanceOrThrow(SystemUtils.loadClass(null, str), getParameters(objArr));
    }

    public static void setField(Object obj, String str, Object obj2) {
        try {
            setFieldOrThrow(obj, str, obj2);
        } catch (Exception e) {
            Log.w(LOG_TAG, "Meet exception when call setField '" + str + "' in " + obj + ", " + e);
        }
    }

    public static void setFieldOrThrow(Object obj, String str, Object obj2) throws IllegalAccessException, NoSuchFieldException {
        Class<?> superclass = obj.getClass();
        Field declaredField = null;
        while (declaredField == null) {
            try {
                declaredField = superclass.getDeclaredField(str);
            } catch (NoSuchFieldException e) {
                superclass = superclass.getSuperclass();
            }
            if (superclass == null) {
                throw new NoSuchFieldException();
            }
        }
        declaredField.setAccessible(true);
        declaredField.set(obj, obj2);
    }
}

File: protocol/src/main/java/com/xiaomi/channel/commonutils/reflect/JavaCalls.kt
package com.xiaomi.channel.commonutils.reflect

import android.util.Log
import com.xiaomi.channel.commonutils.android.SystemUtils
import java.lang.reflect.Constructor
import java.lang.reflect.Field
import java.lang.reflect.InvocationTargetException
import java.lang.reflect.Method

object JavaCalls {
    private const val LOG_TAG = "JavaCalls"

    private val PRIMITIVE_MAP: MutableMap<Class<*>, Class<*>> = hashMapOf(
        Boolean::class.java to Boolean::class.javaPrimitiveType!!,
        Byte::class.java to Byte::class.javaPrimitiveType!!,
        Char::class.java to Char::class.javaPrimitiveType!!,
        Short::class.java to Short::class.javaPrimitiveType!!,
        Int::class.java to Int::class.javaPrimitiveType!!,
        Float::class.java to Float::class.javaPrimitiveType!!,
        Long::class.java to Long::class.javaPrimitiveType!!,
        Double::class.java to Double::class.javaPrimitiveType!!,
        Boolean::class.javaPrimitiveType!! to Boolean::class.javaPrimitiveType!!,
        Byte::class.javaPrimitiveType!! to Byte::class.javaPrimitiveType!!,
        Char::class.javaPrimitiveType!! to Char::class.javaPrimitiveType!!,
        Short::class.javaPrimitiveType!! to Short::class.javaPrimitiveType!!,
        Int::class.javaPrimitiveType!! to Int::class.javaPrimitiveType!!,
        Float::class.javaPrimitiveType!! to Float::class.javaPrimitiveType!!,
        Long::class.javaPrimitiveType!! to Long::class.javaPrimitiveType!!,
        Double::class.javaPrimitiveType!! to Double::class.javaPrimitiveType!!
    )

    class JavaParam<T>(
        @JvmField val clazz: Class<out T>,
        @JvmField val obj: T
    )

    @JvmStatic
    fun callMethod(obj: Any?, name: String, vararg args: Any?): Any? {
        return try {
            if (obj != null) callMethodOrThrow(obj, name, *args) else null
        } catch (e: Exception) {
            Log.w(LOG_TAG, "Meet exception when call Method '$name' in $obj, $e")
            null
        }
    }

    @JvmStatic
    @Throws(
        IllegalAccessException::class, NoSuchMethodException::class,
        SecurityException::class, IllegalArgumentException::class,
        InvocationTargetException::class
    )
    fun callMethodOrThrow(obj: Any?, name: String, vararg args: Any?): Any? {
        if (obj == null) return null
        val paramTypes = getParameterTypes(*args)
        val params = getParameters(*args)
        return getDeclaredMethod(obj.javaClass, name, *paramTypes)
            .invoke(obj, *params)
    }

    @JvmStatic
    fun callStaticMethod(className: String, name: String, vararg args: Any?): Any? {
        return try {
            callStaticMethodOrThrow(SystemUtils.loadClass(null, className), name, *args)
        } catch (e: Exception) {
            Log.w(LOG_TAG, "Meet exception when call Method '$name' in $className, $e")
            null
        }
    }

    @JvmStatic
    @Throws(
        IllegalAccessException::class, NoSuchMethodException::class,
        SecurityException::class, IllegalArgumentException::class,
        InvocationTargetException::class
    )
    fun callStaticMethodOrThrow(cls: Class<*>, name: String, vararg args: Any?): Any {
        val paramTypes = getParameterTypes(*args)
        val params = getParameters(*args)
        return getDeclaredMethod(cls, name, *paramTypes)
            .invoke(null, *params) as Any
    }

    @JvmStatic
    @Throws(
        IllegalAccessException::class, NoSuchMethodException::class,
        SecurityException::class, ClassNotFoundException::class,
        IllegalArgumentException::class, InvocationTargetException::class
    )
    fun callStaticMethodOrThrow(className: String, name: String, vararg args: Any?): Any {
        val paramTypes = getParameterTypes(*args)
        val params = getParameters(*args)
        return getDeclaredMethod(
            SystemUtils.loadClass(null, className),
            name,
            *paramTypes
        ).invoke(null, *params) as Any
    }

    private fun compareClassLists(clsArr: Array<Class<*>>, clsArr2: Array<out Class<*>?>): Boolean {
        if (clsArr2.isEmpty()) {
            return clsArr.isEmpty()
        }
        if (clsArr.size != clsArr2.size) {
            return false
        }
        for (i in clsArr.indices) {
            val cls2 = clsArr2[i]
            if (cls2 != null && !clsArr[i].isAssignableFrom(cls2)) {
                val map = PRIMITIVE_MAP
                if (!map.containsKey(clsArr[i]) || map[clsArr[i]] != map[cls2]) {
                    return false
                }
            }
        }
        return true
    }

    private fun findMethodByName(methodArr: Array<Method>, name: String, paramTypes: Array<out Class<*>?>): Method? {
        for (method in methodArr) {
            if (method.name == name && compareClassLists(method.parameterTypes, paramTypes)) {
                return method
            }
        }
        return null
    }

    @Throws(NoSuchMethodException::class, SecurityException::class)
    private fun getDeclaredMethod(cls: Class<*>, name: String, vararg paramTypes: Class<*>?): Method {
        val method = findMethodByName(cls.declaredMethods, name, paramTypes)
        if (method != null) {
            method.isAccessible = true
            return method
        }
        val superCls = cls.superclass
        if (superCls != null) {
            return getDeclaredMethod(superCls, name, *paramTypes)
        }
        throw NoSuchMethodException()
    }

    private fun getDefaultValue(cls: Class<*>): Any? {
        return when {
            Int::class.java.isAssignableFrom(cls) || Int::class.javaPrimitiveType == cls ||
            Byte::class.java.isAssignableFrom(cls) || Byte::class.javaPrimitiveType == cls ||
            Short::class.java.isAssignableFrom(cls) || Short::class.javaPrimitiveType == cls ||
            Long::class.java.isAssignableFrom(cls) || Long::class.javaPrimitiveType == cls ||
            Double::class.java.isAssignableFrom(cls) || Double::class.javaPrimitiveType == cls ||
            Float::class.java.isAssignableFrom(cls) || Float::class.javaPrimitiveType == cls -> 0
            Boolean::class.java.isAssignableFrom(cls) || Boolean::class.javaPrimitiveType == cls -> false
            Char::class.java.isAssignableFrom(cls) || Char::class.javaPrimitiveType == cls -> 0.toChar()
            else -> null
        }
    }

    @JvmStatic
    fun getField(obj: Any, name: String): Any? {
        return try {
            getFieldOrThrow(obj.javaClass, obj, name)
        } catch (e: Exception) {
            Log.w(LOG_TAG, "Meet exception when call getField '$name' in $obj, $e")
            null
        }
    }

    @Throws(IllegalAccessException::class, NoSuchFieldException::class)
    fun getFieldOrThrow(cls: Class<out Any>, obj: Any?, name: String): Any {
        var superclass: Class<*>? = cls
        var field: Field? = null
        while (field == null) {
            try {
                val declaredField = superclass!!.getDeclaredField(name)
                declaredField.isAccessible = true
                field = declaredField
            } catch (e: NoSuchFieldException) {
                superclass = superclass?.superclass
            }
            if (superclass == null) {
                throw NoSuchFieldException()
            }
        }
        field.isAccessible = true
        return field.get(obj) as Any
    }

    private fun getParameterTypes(vararg args: Any?): Array<Class<*>?> {
        if (args.isEmpty()) return emptyArray()
        return args.map { obj ->
            if (obj is JavaParam<*>) {
                obj.clazz
            } else {
                obj?.javaClass
            }
        }.toTypedArray()
    }

    private fun getParameters(vararg args: Any?): Array<Any?> {
        if (args.isEmpty()) return emptyArray()
        return args.map { obj ->
            if (obj is JavaParam<*>) {
                obj.obj
            } else {
                obj
            }
        }.toTypedArray()
    }

    @JvmStatic
    fun getStaticField(cls: Class<out Any>?, name: String): Any? {
        return try {
            getFieldOrThrow(cls!!, null, name)
        } catch (e: Exception) {
            Log.w(
                LOG_TAG,
                "Meet exception when call getStaticField '$name' in ${cls?.simpleName}, $e"
            )
            null
        }
    }

    @JvmStatic
    fun getStaticField(className: String, name: String): Any? {
        return try {
            getFieldOrThrow(SystemUtils.loadClass(null, className), null, name)
        } catch (e: Exception) {
            Log.w(LOG_TAG, "Meet exception when call getStaticField '$name' in $className, $e")
            null
        }
    }

    @JvmStatic
    fun newEmptyInstance(cls: Class<*>): Any? {
        return try {
            newEmptyInstanceOrThrow(cls)
        } catch (e: Exception) {
            Log.w(LOG_TAG, "Meet exception when make instance as a ${cls.simpleName}, $e")
            null
        }
    }

    @Throws(
        IllegalAccessException::class, InstantiationException::class,
        ClassNotFoundException::class, InvocationTargetException::class
    )
    fun newEmptyInstanceOrThrow(cls: Class<*>): Any {
        val declaredConstructors = cls.declaredConstructors
        if (declaredConstructors.isEmpty()) {
            throw IllegalArgumentException("Can't get even one available constructor for $cls")
        }
        val constructor: Constructor<*> = declaredConstructors[0]
        constructor.isAccessible = true
        val parameterTypes = constructor.parameterTypes
        if (parameterTypes.isEmpty()) {
            return constructor.newInstance()
        }
        val objArr = arrayOfNulls<Any>(parameterTypes.size)
        for (i in parameterTypes.indices) {
            objArr[i] = getDefaultValue(parameterTypes[i])
        }
        return constructor.newInstance(*objArr)
    }

    @JvmStatic
    fun newInstance(cls: Class<*>, vararg args: Any?): Any? {
        return try {
            newInstanceOrThrow(cls, *args)
        } catch (e: Exception) {
            Log.w(LOG_TAG, "Meet exception when make instance as a ${cls.simpleName}, $e")
            null
        }
    }

    @JvmStatic
    fun newInstance(className: String, vararg args: Any?): Any? {
        return try {
            newInstanceOrThrow(className, *args)
        } catch (e: Exception) {
            Log.w(LOG_TAG, "Meet exception when make instance as a $className, $e")
            null
        }
    }

    @Throws(
        IllegalAccessException::class, NoSuchMethodException::class,
        InstantiationException::class, SecurityException::class,
        IllegalArgumentException::class, InvocationTargetException::class
    )
    fun newInstanceOrThrow(cls: Class<*>, vararg args: Any?): Any {
        val paramTypes = getParameterTypes(*args)
        val params = getParameters(*args)
        return cls.getConstructor(*paramTypes).newInstance(*params)
    }

    @Throws(
        IllegalAccessException::class, NoSuchMethodException::class,
        InstantiationException::class, SecurityException::class,
        ClassNotFoundException::class, IllegalArgumentException::class,
        InvocationTargetException::class
    )
    fun newInstanceOrThrow(className: String, vararg args: Any?): Any {
        val params = getParameters(*args)
        return newInstanceOrThrow(
            SystemUtils.loadClass(null, className),
            *params
        )
    }

    @JvmStatic
    fun setField(obj: Any, name: String, value: Any?) {
        try {
            setFieldOrThrow(obj, name, value)
        } catch (e: Exception) {
            Log.w(LOG_TAG, "Meet exception when call setField '$name' in $obj, $e")
        }
    }

    @Throws(IllegalAccessException::class, NoSuchFieldException::class)
    fun setFieldOrThrow(obj: Any, name: String, value: Any?) {
        var superclass: Class<*>? = obj.javaClass
        var declaredField: Field? = null
        while (declaredField == null) {
            try {
                declaredField = superclass!!.getDeclaredField(name)
            } catch (e: NoSuchFieldException) {
                superclass = superclass?.superclass
            }
            if (superclass == null) {
                throw NoSuchFieldException()
            }
        }
        declaredField.isAccessible = true
        declaredField.set(obj, value)
    }
}
